package/src/main/java/com/microsoft/azure/cosmos/cassandra/CosmosDBLoadBalancingPolicy.java

```diff
@@ -0,0 +1,359 @@
+        CopyOnWriteArrayList<Host> readLocalDCAddresses = new CopyOnWriteArrayList<Host>();
+        CopyOnWriteArrayList<Host> writeLocalDCAddresses = new CopyOnWriteArrayList<Host>();
+        CopyOnWriteArrayList<Host> remoteDCAddresses = new CopyOnWriteArrayList<Host>();
+
+        List<InetAddress> dnsLookupAddresses = new ArrayList<InetAddress>();
+        if (!globalContactPoint.isEmpty()) {
+            dnsLookupAddresses = Arrays.asList(getLocalAddresses());
+        }
+
+        for (Host host : hosts) {
+            if (!this.readDC.isEmpty() && host.getDatacenter().equals(readDC)) {
+                readLocalDCAddresses.add(host);
+            }
+
+            if ((!this.writeDC.isEmpty() && host.getDatacenter().equals(writeDC))
+                    || dnsLookupAddresses.contains(host.getAddress())) {
+                writeLocalDCAddresses.add(host);
+            }
+            else {
+                remoteDCAddresses.add(host);
+            }
+        }
+
+        this.readLocalDCHosts = readLocalDCAddresses;
+        this.writeLocalDCHosts = writeLocalDCAddresses;
+        this.remoteDCHosts = remoteDCAddresses;
+        this.index.set(new Random().nextInt(Math.max(hosts.size(), 1)));
+    }
+
+    /**
+     * Return the HostDistance for the provided host.
+     *
+     * <p>This policy consider the nodes for the writeDC and the default write region at distance {@code LOCAL}.
+     *
+     * @param host the host of which to return the distance of.
+     * @return the HostDistance to {@code host}.
+     */
+    @Override
+    public HostDistance distance(Host host) {
+        if (!this.writeDC.isEmpty()) {
+            if (host.getDatacenter().equals(this.writeDC)) {
+                return HostDistance.LOCAL;
+            }
+        } else if (Arrays.asList(getLocalAddresses()).contains(host.getAddress())) {
+            return HostDistance.LOCAL;
+        }
+
+        return HostDistance.REMOTE;
+    }
+
+    /**
+     * Returns the hosts to use for a new query.
+     *
+     * <p>For read requests, the returned plan will always try each known host in the readDC first.
+     * if none of the host is reachable, it will try all other hosts.
+     * For write and all other requests, the returned plan will always try each known host in the writeDC or the
+     * default write region first. If none of the host is reachable, it will try all other hosts.
+     *
+     * @param loggedKeyspace the keyspace currently logged in on for this query.
+     * @param statement the query for which to build the plan.
+     * @return a new query plan, i.e. an iterator indicating which host to try first for querying,
+     *     which one to use as failover, etc...
+     */
+    @Override
+    public Iterator<Host> newQueryPlan(String loggedKeyspace, final Statement statement) {
+        refreshHostsIfDnsExpired();
+
+        final List<Host> readHosts = cloneList(this.readLocalDCHosts);
```

Comment:
Can we avoid a clone here?  Isnt this expensive to clone an array for every request?

Reply:
This is for the concurrency handling. The list can be modified for the duration of computeNext(). I followed the pattern here: https://github.com/tfredrich/cassandra-java-driver/blob/master/driver-core/src/main/java/com/datastax/driver/core/policies/DCAwareRoundRobinPolicy.java